File: src/lib/sm2.ts
export type ReviewQuality = "again" | "know";

export type Sm2State = {
  easeFactor: number;
  intervalDays: number;
  repetitions: number;
};

const MIN_EASE = 1.3;

/** SM-2 quality: again = 1, know = 5 */
function toQuality(quality: ReviewQuality): number {
  return quality === "know" ? 5 : 1;
}

export function sm2Next(
  state: Sm2State,
  quality: ReviewQuality,
  today: string,
): Sm2State & { nextReviewDate: string } {
  const q = toQuality(quality);
  let { easeFactor, intervalDays, repetitions } = state;

  if (q < 3) {
    repetitions = 0;
    intervalDays = 1;
  } else {
    if (repetitions === 0) {
      intervalDays = 1;
    } else if (repetitions === 1) {
      intervalDays = 6;
    } else {
      intervalDays = Math.max(1, Math.round(intervalDays * easeFactor));
    }
    repetitions += 1;
  }

  easeFactor = Math.max(
    MIN_EASE,
    easeFactor + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02)),
  );

  const next = new Date(`${today}T00:00:00.000Z`);
  next.setUTCDate(next.getUTCDate() + intervalDays);

  return {
    easeFactor,
    intervalDays,
    repetitions,
    nextReviewDate: next.toISOString().slice(0, 10),
  };
}

export const DEFAULT_SM2_STATE: Sm2State = {
  easeFactor: 2.5,
  intervalDays: 0,
  repetitions: 0,
};

File: backend/polskiflow/domain/sm2.py
"""SM-2 scheduling rules shared by future web and background jobs."""

from dataclasses import dataclass
from datetime import date, timedelta
from math import floor
from typing import Literal

ReviewQuality = Literal["again", "hard", "good", "easy", "know"]
MIN_EASE_FACTOR = 1.3
QUALITY_SCORES: dict[ReviewQuality, int] = {
    "again": 1,
    "hard": 3,
    "good": 4,
    "easy": 5,
    # Compatibility with the first two-button flashcard MVP.
    "know": 5,
}


@dataclass(frozen=True)
class Sm2State:
    ease_factor: float
    interval_days: int
    repetitions: int


@dataclass(frozen=True)
class Sm2Result(Sm2State):
    next_review_date: date


DEFAULT_SM2_STATE = Sm2State(
    ease_factor=2.5,
    interval_days=0,
    repetitions=0,
)


def sm2_next(state: Sm2State, quality: ReviewQuality, today: date) -> Sm2Result:
    """Return the next immutable review state using the SM-2 quality scale."""

    if quality not in QUALITY_SCORES:
        raise ValueError("unknown review quality")

    score = QUALITY_SCORES[quality]
    repetitions = state.repetitions
    interval_days = state.interval_days

    if score < 3:
        repetitions = 0
        interval_days = 1
    else:
        if repetitions == 0:
            interval_days = 1
        elif repetitions == 1:
            interval_days = 6
        else:
            # JavaScript Math.round rounds positive halves upward; Python's
            # round uses bankers' rounding, so preserve the MVP behaviour.
            interval_days = max(1, floor(interval_days * state.ease_factor + 0.5))
        repetitions += 1

    ease_factor = max(
        MIN_EASE_FACTOR,
        state.ease_factor
        + (0.1 - (5 - score) * (0.08 + (5 - score) * 0.02)),
    )

    return Sm2Result(
        ease_factor=ease_factor,
        interval_days=interval_days,
        repetitions=repetitions,
        next_review_date=today + timedelta(days=interval_days),
    )
